mutiny-core/src/storage.rs

```diff
@@ -653,6 +663,99 @@ impl MutinyStorage for () {
     }
 }
 
+fn payment_key(inbound: bool, payment_hash: &[u8; 32]) -> String {
+    if inbound {
+        format!(
+            "{}{}",
+            PAYMENT_INBOUND_PREFIX_KEY,
+            payment_hash.to_hex().as_str()
+        )
+    } else {
+        format!(
+            "{}{}",
+            PAYMENT_OUTBOUND_PREFIX_KEY,
+            payment_hash.to_hex().as_str()
+        )
+    }
+}
+
+pub(crate) fn persist_payment_info<S: MutinyStorage>(
+    storage: S,
+    payment_hash: &[u8; 32],
+    payment_info: &PaymentInfo,
+    inbound: bool,
+) -> std::io::Result<()> {
+    let key = payment_key(inbound, payment_hash);
+    storage
+        .set_data(key, payment_info, None)
+        .map_err(std::io::Error::other)
+}
+
+pub(crate) fn get_payment_info<S: MutinyStorage>(
+    storage: S,
+    payment_hash: &bitcoin::hashes::sha256::Hash,
+    logger: &MutinyLogger,
+) -> Result<(PaymentInfo, bool), MutinyError> {
+    // try inbound first
+    if let Some(payment_info) =
+        read_payment_info(storage.clone(), payment_hash.as_inner(), true, logger)
+    {
+        return Ok((payment_info, true));
+    }
+
+    // if no inbound check outbound
+    match read_payment_info(storage.clone(), payment_hash.as_inner(), false, logger) {
+        Some(payment_info) => Ok((payment_info, false)),
+        None => Err(MutinyError::NotFound),
+    }
+}
+
+pub(crate) fn read_payment_info<S: MutinyStorage>(
+    storage: S,
+    payment_hash: &[u8; 32],
+    inbound: bool,
+    logger: &MutinyLogger,
+) -> Option<PaymentInfo> {
+    let key = payment_key(inbound, payment_hash);
+    log_trace!(logger, "Trace: checking payment key: {key}");
+    match storage.get_data(&key).transpose() {
+        Some(Ok(v)) => Some(v),
+        _ => {
+            // To scan for the old format that had `_{node_id}` at the end
+            if let Ok(map) = storage.scan(&key, None) {
+                map.into_values().next()
+            } else {
+                None
+            }
+        }
+    }
+}
+
+pub(crate) fn list_payment_info<S: MutinyStorage>(
+    storage: S,
+    inbound: bool,
+) -> Result<Vec<(PaymentHash, PaymentInfo)>, MutinyError> {
+    let prefix = match inbound {
+        true => PAYMENT_INBOUND_PREFIX_KEY,
+        false => PAYMENT_OUTBOUND_PREFIX_KEY,
+    };
+    let map: HashMap<String, PaymentInfo> = storage.scan(prefix, None)?;
+
+    // convert keys to PaymentHash
+    Ok(map
+        .into_iter()
+        .map(|(key, value)| {
+            let payment_hash_str = key
+                .trim_start_matches(prefix)
+                .splitn(2, '_') // To support the old format that had `_{node_id}` at the end
```

Comment:
does this handle if there is no `_` suffix?

Reply:
Yes, it's a max split of 2 so if there's none then it's just the whole string as the first vector. Always guaranteed to have 1 entry so the `[0]` is safe.